File: 5회차/1611_Minimum_One_Bit_Operations_to_Make_Integers_Zero/1611_Minimum_One_Bit_Operations_to_Make_Integers_Zero.cpp
#include <iostream>

class Solution {
public:
    int minimumOneBitOperations(int n) {
        if (n == 0)
            return 0;

        // Find the position of the most significant bit (MSB)
        int k = 0;
        while ((1 << k) <= n)
            k++;
        k--;

        return (1 << (k + 1)) - 1 - minimumOneBitOperations(n ^ (1 << k));
    }
};

// Example usage
int main() {
    Solution solution;
    int n1 = 3;
    int result1 = solution.minimumOneBitOperations(n1);
    std::cout << "Minimum operations to reduce " << n1 << " to 0: " << result1 << std::endl; // Output: 2

    int n2 = 6;
    int result2 = solution.minimumOneBitOperations(n2);
    std::cout << "Minimum operations to reduce " << n2 << " to 0: " << result2 << std::endl; // Output: 4

    return 0;
}

File: 5회차/1611_Minimum_One_Bit_Operations_to_Make_Integers_Zero/1611_Minimum_One_Bit_Operations_to_Make_Integers_Zero.py
class Solution:
    def minimumOneBitOperations(self, n: int) -> int:
        if n == 0:
            return 0
        # Find the position of the most significant bit (MSB)
        k = 0
        while 2**k <= n:
            k += 1
        k -= 1

        return 2**(k+1)-1 - self.minimumOneBitOperations(2**k ^ n)


# Example usage:
if __name__ == "__main__":
    solution = Solution()
    print(solution.minimumOneBitOperations(3))  # Output: 2
    print(solution.minimumOneBitOperations(6))  # Output: 4